### PR TITLE
fix #4160 by detecting ASP.NET Core Servers

### LINKED ISSUE #4160: Add useful error message when connecting to ASP.NET Core SignalR Server
The ASP.NET SignalR Client is not compatible with the ASP.NET Core SignalR server. We should detect this in the clients and add a useful error message.

### PR DESCRIPTION
**Pull request by anurse**
fixes #4160

This change adds logic to the JS and .NET clients to detect ASP.NET Core servers using the 'availableTransports' property in the negotiate response, which is unique to ASP.NET Core

### CODE REVIEW

File: src/Microsoft.AspNet.SignalR.Client/Resources.resx [hunk 1]
```diff
@@ -192,4 +192,7 @@
   <data name="Error_ErrorFromServer" xml:space="preserve">
     <value>Error message received from the server: '{0}'.</value>
   </data>
+  <data name="Error_AspNetCoreServerDetected" xml:space="preserve">
+    <value>Detected a connection attempt to an ASP.NET SignalR Server. This client only supports connecting to an ASP.NET Core SignalR Server. See https://aka.ms/signalr-core-differences for details.</value>
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Microsoft.AspNet.SignalR.Client/Resources.resx]: Yah.

[the remark this answers is not in the document]
Comment (anurse) [reply src/Microsoft.AspNet.SignalR.Client/Resources.resx]: Yep, just bad copy-pasta

File: test/Microsoft.AspNet.SignalR.Tests.Common/App_Start/Initializer.cs [hunk 2]
```diff
@@ -403,23 +427,6 @@ private static void RegisterSignalREndpoints(IAppBuilder app, IDependencyResolve
             {
                 map.MapSignalR();
             });
-
-            // Redirectors:
-
-            // Valid redirect chain
-            // Overload detection doesn't like it when we use this as an extension method
-            // We *intentionally* use paths that do NOT end in a trailing '/' in some places as the client needs to support both
-            AppBuilderUseExtensions.Use(app, CreateRedirector("/redirect", "/redirect2"));
```
Comment (anurse) [inline test/Microsoft.AspNet.SignalR.Tests.Common/App_Start/Initializer.cs]: These were duplicated due to a bad merge earlier

**Comment by anurse** [comment]
:up: :date: fixed the order

File: src/Microsoft.AspNet.SignalR.JS/jquery.signalR.core.js [hunk 3]
```diff
@@ -720,6 +721,14 @@
                             return;
                         }
 
+                        // Check if the server is an ASP.NET Core app
+                        if (res.availableTransports) {
+                            protocolError = signalR._.error(resources.aspnetCoreSignalrServer);
```
Comment (mikaelm12) [inline src/Microsoft.AspNet.SignalR.JS/jquery.signalR.core.js]: ._. ?

File: test/Microsoft.AspNet.SignalR.FunctionalTests/Client/ConnectionFacts.cs [hunk 4]
```diff
@@ -469,5 +469,27 @@ public async Task ConnectionFunctionsCorrectlyAfterCallingStartMutlipleTimes(Hos
...
+        {
+            using (var host = CreateHost(hostType, transportType))
+            {
+                host.Initialize();
+
+                using (var connection = CreateConnection(host, "/aspnetcore-signalr"))
+                {
+                    var ex = await Assert.ThrowsAsync<InvalidOperationException>(() => connection.Start(host.TransportFactory())).OrTimeout();
+                    Assert.Equal(
+                        "Detected a connection attempt to an ASP.NET Core SignalR Server. This client only supports connecting to an ASP.NET SignalR Server. See https://aka.ms/signalr-core-differences for details.",
```
Comment (mikaelm12) [inline test/Microsoft.AspNet.SignalR.FunctionalTests/Client/ConnectionFacts.cs]: nit: Can you just use the `Resources.Error_AspNetCoreServerDetected` property here?

Comment (anurse) [reply test/Microsoft.AspNet.SignalR.FunctionalTests/Client/ConnectionFacts.cs]: It's not accessible from here and I think it's worth adding IVT just for that.

File: src/Microsoft.AspNet.SignalR.JS/jquery.signalR.core.js (hunk 3 above)
Comment (anurse) [reply src/Microsoft.AspNet.SignalR.JS/jquery.signalR.core.js]: That's where the function is defined.

**Pull request merged**